openverse_catalog/dags/providers/provider_dag_factory.py

```diff
@@ -276,82 +256,134 @@ def create_provider_api_workflow(
-    dag_id: str,
-    main_function: Callable,
-    reingestion_day_list_list: list[list[int]],
-    start_date: datetime = datetime(1970, 1, 1),
-    max_active_runs: int = 1,
-    max_active_tasks: int = 1,
-    default_args: Optional[Dict] = None,
-    dagrun_timeout: timedelta = timedelta(hours=23),
-    ingestion_task_timeout: timedelta = timedelta(hours=2),
+    conf: ProviderReingestionWorkflow, reingestion_day_list_list: list[list[int]]
 ):
     """
-    Given a `main_function` and `reingestion_day_list_list`, this
-    factory method instantiates a DAG that will run the given
-    `main_function`, parameterized by a number of dates, whose
-    calculation is described below.
+    Given a `conf` object and `reingestion_day_list_list`, this
+    factory method instantiates a DAG that will run ingestion using the
+    given configuration, parameterized by a number of dates calculated
+    using the reingestion day list.
 
     Required Arguments:
 
-    dag_id:                     string giving a unique id of the DAG to
-                                be created.
-    main_function:              python function to be run. The
-                                function must take a single parameter
-                                (date) which will be a string of the
-                                form 'YYYY-MM-DD'.
+    conf:                       ProviderReingestionWorkflow configuration
+                                object used to configure the ingestion tasks.
     reingestion_day_list_list:  list of lists of integers. It gives the
                                 set of days before the current execution
                                 date of the DAG for which the
                                 `main_function` should be run, and
                                 describes how the calls to the function
                                 should be prioritized.
+    """
+    default_args = {**DAG_DEFAULT_ARGS, **(conf.default_args or {})}
+    dag = DAG(
+        dag_id=conf.dag_id,
+        default_args={**default_args, "start_date": conf.start_date},
+        max_active_tasks=conf.max_active_tasks,
+        max_active_runs=conf.max_active_runs,
+        dagrun_timeout=conf.dagrun_timeout,
+        schedule_interval=conf.schedule_string,
+        start_date=conf.start_date,
+        catchup=False,
+        doc_md=conf.doc_md,
+        tags=["provider-reingestion"],
+        render_template_as_native_obj=True,
+        user_defined_macros={"date_partition_for_prefix": date_partition_for_prefix},
+    )
+    with dag:
+        # Generate a list of lists of ingestion TaskGroups for each day of reingestion.
+        ingest_operator_list_list, ingestion_metrics = _build_ingest_operator_list_list(
+            reingestion_day_list_list, conf
+        )
+
+        # For each 'level', make a wait task that waits for all of the reingestion taks
+        # at that level to complete.
+        for i in range(len(ingest_operator_list_list) - 1):
```

Comment:
I would love to do this, but it would take some refactoring. Currently it's possible to define a list of day-shifts that includes, say, "daily" runs and "one_month" runs but skips the "weekly" and "fortnightly". I'll think about how we could do this cleanly 🤔

Reply:
Woah, okay that's interesting. Maybe we could just expand this out to `level_0` or `tier_0` for now then?